README.md
# Fast Polylines

Implementation of the [Google polyline algorithm][algorithm].

Up to 60x faster than the [polyline package][polyline].

## Benchmark

`make bench` made on a Apple M1 computer.


### encode

```
fast_polyline:	10.606ms
polyline:	613.5ms
fast_polyline is 57.8 times faster.
```

### decode

```
fast_polyline:	23.447ms
polyline:	425.96ms
fast_polyline is 18.2 times faster.
```


## Install

```bash
pip install fast_polyline
```

or in your `requirements.txt`:

```ruby
fast_polyline~=0.2.0
```

## Usage

```python
import fast_polyline

fast_polyline.encode([[38.5, -120.2], [40.7, -120.95], [43.252, -126.453]])
# => "_p~iF~ps|U_ulLnnqC_mqNvxq`@"

fast_polyline.decode("_p~iF~ps|U_ulLnnqC_mqNvxq`@")
# => [[38.5, -120.2], [40.7, -120.95], [43.252, -126.453]]
```

## Advanced usage

**Use a different precision**

Default precision is `5` decimals, to use a precision of `6`:
```ruby
fast_polyline.encode([[38.5, -120.2], [40.7, -120.95], [43.252, -126.453]], 6)
# => "_izlhA~rlgdF_{geC~ywl@_kwzCn`{nI"

fast_polyline.decode("_izlhA~rlgdF_{geC~ywl@_kwzCn`{nI", 6)
# => [[38.5, -120.2], [40.7, -120.95], [43.252, -126.453]]
```
The precision max is `13`. And I advanced against using it
since [it is already way too precise][xkcd].

`Encoder` and `Decoder` modules are deprec

## Contributing

```bash
git clone git@github.com:BuonOmo/fast-polyline.py
cd fast-polyline.py
make # show all commands you might need
# Implement a feature, resolve a bug...
make lint
make test
git commit "feat: respect conventional commits"
# Make a PR
```

[And here's the Python C ABI doc.][python-c]

## License

Please see the LICENSE file.

[algorithm]: https://code.google.com/apis/maps/documentation/utilities/polylinealgorithm.html
[polyline]: https://github.com/frederickjansen/polyline
[xkcd]: https://xkcd.com/2170/
[python-c]: https://docs.python.org/3/c-api/index.html


CHANGELOG.md
## v0.2.0 (2023-07-03)

### Feat

- add tests

### Fix

- **ext**: clarify errors

### Refactor

- reorganize makefile
- rename C file
- Set doc and precision in python
- **setup**: move setup to pyproject.toml

## 0.1.0 (2023-06-25)

### Feat

- **ext**: set error messages
- **bench**: add benchmark
- **ext**: add core logic from the ruby gem
- **Makefile**: add clangd configuration
- initialize repo

### Fix

- **encode**: out of the loop, dummy

### Refactor

- **bench**: formatting and output
- **name**: switch to fast_polyline


tests/test_fast_polyline.py
from random import uniform, randint
import time

import fast_polyline as polyline


def test_decode_multiple_points():
    d = polyline.decode('gu`wFnfys@???nKgE??gE?????oK????fE??fE')
    assert d == [
        (40.641, -8.654),
        (40.641, -8.654),
        (40.641, -8.656),
        (40.642, -8.656),
        (40.642, -8.655),
        (40.642, -8.655),
        (40.642, -8.655),
        (40.642, -8.653),
        (40.642, -8.653),
        (40.642, -8.653),
        (40.641, -8.653),
        (40.641, -8.654)
    ]


def test_decode_multiple_points_precision():
    d = polyline.decode('_epolA~ieoOnF??~{Bo}@??o}@?????_|B????n}@??n}@', 6)
    assert d == [
        (40.64112, -8.654),
        (40.641, -8.654),
        (40.641, -8.656),
        (40.642, -8.656),
        (40.642, -8.655),
        (40.642, -8.655),
        (40.642, -8.655),
        (40.642, -8.653),
        (40.642, -8.653),
        (40.642, -8.653),
        (40.641, -8.653),
        (40.641, -8.654)
    ]

def test_decode_official_example():
    d = polyline.decode('_p~iF~ps|U_ulLnnqC_mqNvxq`@')
    assert d == [
        (38.500, -120.200),
        (40.700, -120.950),
        (43.252, -126.453)
    ]

def test_decode_official_example_precision():
    d = polyline.decode('_izlhA~rlgdF_{geC~ywl@_kwzCn`{nI', 6)
    assert d == [
        (38.500, -120.200),
        (40.700, -120.950),
        (43.252, -126.453)
    ]


def test_decode_single_point():
    d = polyline.decode('gu`wFf`ys@')
    assert d == [
        (40.641, -8.653)
    ]


def test_decode_single_point_precision():
    d = polyline.decode('o}oolAnkcoO', 6)
    assert d == [
        (40.641, -8.653)
    ]


def test_encode_multiple_points():
    e = polyline.encode([
        (40.641, -8.654),
        (40.641, -8.654),
        (40.641, -8.656),
        (40.642, -8.656),
        (40.642, -8.655),
        (40.642, -8.655),
        (40.642, -8.655),
        (40.642, -8.653),
        (40.642, -8.653),
        (40.642, -8.653),
        (40.641, -8.653),
        (40.641, -8.654)
    ])
    assert e == 'gu`wFnfys@???nKgE??gE?????oK????fE??fE'


def test_encode_multiple_points_precision():
    e = polyline.encode([
        (40.64112345, -8.654),
        (40.641, -8.654),
        (40.641, -8.656),
        (40.642, -8.656),
        (40.642, -8.655),
        (40.642, -8.655),
        (40.642, -8.655),
        (40.642, -8.653),
        (40.642, -8.653),
        (40.642, -8.653),
        (40.641, -8.653),
        (40.641, -8.654)
    ], 6)
    assert e == 'eepolA~ieoOtF??~{Bo}@??o}@?????_|B????n}@??n}@'

def test_encode_official_example():
    e = polyline.encode([
        (38.500, -120.200),
        (40.700, -120.950),
        (43.252, -126.453)
    ])
    assert e == '_p~iF~ps|U_ulLnnqC_mqNvxq`@'


def test_encode_official_example_precision():
    e = polyline.encode([
        (38.500, -120.200),
        (40.700, -120.950),
        (43.252, -126.453)
    ], 6)
    assert e == '_izlhA~rlgdF_{geC~ywl@_kwzCn`{nI'


def test_encode_single_point():
    e = polyline.encode([
        (40.64155, -8.65344)
    ])
    assert e == 'ux`wF~bys@'

    e = polyline.encode([
        (40.641552, -8.653441)
    ])
    assert e == 'ux`wF~bys@'


def test_encode_single_point_rounding():
    e = polyline.encode([
        (0, 0.000006),
        (0, 0.000002)
    ])
    assert e == '?A?@'


def test_rounding_py3_match_py2():
    e = polyline.encode([
        (36.05322, -112.084004),
        (36.053573, -112.083914),
        (36.053845, -112.083965)])
    assert e == 'ss`{E~kbkTeAQw@J'


def test_encode_single_point_precision():
    e = polyline.encode([
        (40.641123, -8.653321)
    ], 6)
    assert e == 'eepolAp_doO'

    e = polyline.encode([
        (40.6411233123, -8.6533214234)
    ], 6)
    assert e == 'eepolAp_doO'


def test_a_variety_of_precisions():
    """uses a generator to create a variety of lat-lon's across the global
        and tests a range of precision settings from 2 to 9"""

    def generator():
        while True:
            coords = []
            for i in range(2, randint(4, 10)):
                lat, lon = uniform(-180.0, 180.0), uniform(-180.0, 180.0)
                coords.append((lat, lon))
            yield coords

    patience = 0.5  # seconds.
    waypoints, okays = 0, 0

    g = generator()
    start = time.time()
    while time.time() < start + patience:
        precision = randint(2, 9)
        wp = next(g)
        waypoints += len(wp)
        poly = polyline.encode(wp, precision)
        wp2 = polyline.decode(poly, precision)
        if wp == wp2:
            okays += len(wp2)
        else:
            for idx, _ in enumerate(wp):
                dx, dy = abs(wp[idx][0] - wp2[idx][0]), abs(wp[idx][1] - wp2[idx][1])
                if dx > 10 ** -(precision - 1) or dy > 10 ** -(precision - 1):
                    print(f"idx={idx}, dx={dx}, dy={dy}")
                else:
                    okays += 1

    assert okays == waypoints
    print(
        f"encoded and decoded {100 * okays / float(waypoints):.2f}% correctly for {waypoints} "
        f"waypoints @ {round(waypoints / patience, 0)} wp/sec")


src/polyline.c
#define PY_SSIZE_T_CLEAN
#include <Python.h>

/* # Resources
 *
 * [Python C-API reference](https://docs.python.org/3/c-api/index.html).
 */

/*
 * TODO: readme (.rst???)
 * TODO: inline documentation, however it works in C Python.
 * TODO: test
 *   - check that error messages are set.
 * TODO: offer merging with the existing polyline project
 *   https://github.com/frederickjansen/polyline/issues/14
 * TODO: choose final naming (- or _ depending on how they do in python)
 */

// An encoded number can have at most _precision_ characters. However,
// it seems like we have a fuzzy behavior on low precisions. Hence a guard
// for those lower precisions.
#define MAX_ENCODED_CHUNKS(precision) (precision < 5 ? 5 : precision)

#define DEFAULT_PRECISION 5

// Already smaller than a sand grain. https://xkcd.com/2170/
// In fact, at 15 precision, we can be sure there will be precision loss
// in the C/Ruby value conversion. And 14 precision may be an edge case
// since we work with signed values.
#define MAX_PRECISION 13

#ifdef DEBUG
#define dbg(...) printf(__VA_ARGS__)
#else
#define dbg(...)
#endif

typedef unsigned int uint;

// Speed up the code a bit by not calling pow(10.0, precision) since it is
// always 10^precision. Convert that into a lookup instead of doing the
// calculation.
// The precision parameter is already checked in +_get_precision()+ to ensure
// that it is within the range.
static inline double _fast_pow10(uint precision) {
	const double lookup[MAX_PRECISION + 1] = {
	    1.0,    10.0,   1.0e+2, 1.0e+3,  1.0e+4,  1.0e+5,  1.0e+6,
	    1.0e+7, 1.0e+8, 1.0e+9, 1.0e+10, 1.0e+11, 1.0e+12, 1.0e+13};
	return lookup[precision];
}

// Ensure that precision is between 0 and 13, set error string
// otherwise.
static inline int _check_precision(int precision) {
	if (precision > MAX_PRECISION) {
		PyErr_SetString(PyExc_ValueError,
		                "precision too high (https://xkcd.com/2170/)");
		return 0;
	}
	if (precision < 0) {
		PyErr_SetString(PyExc_ValueError,
		                "negative precision doesn't make sense");
		return 0;
	}
	return 1;
}

// Wrapper around PyArg_ParseTuple that sets an error message.
static inline int _check_args(PyObject *args, const char *fmt, ...) {
	// Passing through variadic arguments is not that easy in C.
	// Here's some information on the topic:
	// https://c-faq.com/varargs/handoff.html
	va_list argp;
	va_start(argp, fmt);
	int ret = PyArg_VaParse(args, fmt, argp);
	va_end(argp);
	if (!ret) PyErr_SetString(PyExc_TypeError, "invalid arguments");
	return ret;
}

// polyline.decode(polyline, precision=5)
static PyObject *polyline_decode(PyObject *self, PyObject *args) {
	const char *polyline;
	int precision = DEFAULT_PRECISION;
	if (!_check_args(args, "s|i", &polyline, &precision)) return NULL;
	if (!_check_precision(precision)) return NULL;
	double precision_value = _fast_pow10(precision);
	PyObject *ary = PyList_New(0);
	// Helps keeping track of whether we are computing lat (0) or lng (1).
	uint8_t index = 0;
	size_t shift = 0;
	int64_t delta = 0;
	PyObject *pair[2];
	double latlng[2] = {0.0, 0.0};
	// Loops until end of string nul character is encountered.
	while (*polyline) {
		int64_t chunk = *polyline++;

		if (chunk < 63 || chunk > 126) {
			PyErr_Format(PyExc_ValueError, "invalid character '%c'",
			             (char)chunk);
			return NULL;  // TODO: ensure that the objects we created will be
			              // gc'd
		}

		chunk -= 63;
		delta = delta | ((chunk & ~0x20) << shift);
		shift += 5;

		if (!(chunk & 0x20)) {
			delta = (delta & 1) ? ~(delta >> 1) : (delta >> 1);
			latlng[index] += delta;
			pair[index] =
			    PyFloat_FromDouble((double)latlng[index] / precision_value);
			// When both coordinates are parsed, we can push those to the result
			// ary.
			if (index) PyList_Append(ary, PyTuple_Pack(2, pair[0], pair[1]));
			// Reinitilize since we are done for current coordinate.
			index = 1 - index;
			delta = 0;
			shift = 0;
		}
	}
	return ary;
}

static inline uint8_t _polyline_encode_number(char *chunks, int64_t number) {
	dbg("_polyline_encode_number(\"%s\", %lli)\n", chunks, number);
	number = number < 0 ? ~(number << 1) : (number << 1);
	uint8_t i = 0;
	while (number >= 0x20) {
		uint8_t chunk = number & 0x1f;
		chunks[i++] = (0x20 | chunk) + 63;
		number = number >> 5;
	}
	chunks[i++] = number + 63;
	dbg("%u encoded chunks\n", i);
	dbg("chunks: '%s'\n", chunks);
	dbg("/_polyline_encode_number\n");
	return i;
}

// polyline.encode(points, precision=5)
static PyObject *polyline_encode(PyObject *self, PyObject *args) {
	PyObject *ary;
	// Dynamically checked for [0; 13] inclusion.
	int precision = DEFAULT_PRECISION;
	if (!_check_args(args, "O|i", &ary, &precision)) return NULL;
	if (!_check_precision(precision)) return NULL;
	double precision_value = _fast_pow10(precision);
	if (!PyList_Check(ary)) {
		PyErr_SetString(PyExc_TypeError, "points must be a list");
		return NULL;
	}

	size_t len = PyList_Size(ary);
	uint64_t i;
	PyObject *current_pair;
	int64_t prev_pair[2] = {0, 0};

	char *chunks =
	    malloc(MAX_ENCODED_CHUNKS(precision) * 2 * len * sizeof(char));
	size_t chunks_index = 0;
	dbg("ary.len: %lu\n", len);
	for (i = 0; i < len; i++) {
		current_pair = PyList_GetItem(ary, i);
		uint8_t j;
		if (PyList_Check(current_pair))
			current_pair = PyList_AsTuple(current_pair);
		if (!PyTuple_Check(current_pair)) goto points_error;
		if (PyTuple_Size(current_pair) != 2) goto points_error;

		for (j = 0; j < 2; j++) {
			PyObject *current_coord = PyTuple_GetItem(current_pair, j);
			if (!PyFloat_Check(current_coord)) goto points_error;

			double coord = PyFloat_AsDouble(current_coord);
			if (-180.0 > coord || coord > 180.0) goto points_error;

			int64_t rounded_value = round(coord * precision_value);
			int64_t delta = rounded_value - prev_pair[j];
			prev_pair[j] = rounded_value;
			// We pass a pointer to the current chunk that need to be filled.
			// Doing so avoid having to copy the string every single iteration.
			chunks_index += _polyline_encode_number(
			    chunks + chunks_index * sizeof(char), delta);
		}
	}

	PyObject *polyline = PyUnicode_FromStringAndSize(chunks, chunks_index);
	free(chunks);
	return polyline;

points_error:
	PyErr_SetString(PyExc_TypeError,
	                "points must be a list of (lat, lng) pairs");
	free(chunks);
	return NULL;
}

static PyMethodDef methods[] = {
    {"decode", polyline_decode, METH_VARARGS, "Decode a polyline"},
    {"encode", polyline_encode, METH_VARARGS, "Encode a polyline"},
    {NULL, NULL, 0, NULL}};

static struct PyModuleDef module = {PyModuleDef_HEAD_INIT, "fast_polyline",
                                    "Encode and decode polylines", -1, methods};
PyMODINIT_FUNC PyInit_fast_polyline(void) { return PyModule_Create(&module); }


perf/benchmark.py
"""
Benchmark to compare with existing libraries.
"""
import sys
import timeit

import polyline

import fast_polyline

POLYLINE = "{leiHkmjM}z@vcE~JlKx@b~A~Jpy@vv@bhArLpVxFta@sCjEbL~c@rEn{@_HzXj" \
           "EdUg@rfCbHbSp\\fKpQzSg@|fCxTfvAzZduEyC|z@al@l~BcOnzBr@jUfuDfgFf" \
           "xAbzA~HxUfKcE_B}YbEyWti@_rAr`AhNte@x]dLyk@|Vyd@"
POINTS = [
	(48.85726, 2.35238), (48.86685, 2.3209), (48.86493, 2.31891),
	(48.86464, 2.30369), (48.86272, 2.29432), (48.8538, 2.28262),
	(48.85162, 2.27885), (48.85037, 2.2733), (48.85111, 2.27228),
	(48.84901, 2.26636), (48.84795, 2.25668), (48.84939, 2.25254),
	(48.84837, 2.24899), (48.84857, 2.22729), (48.84711, 2.22407),
	(48.84238, 2.22211), (48.83941, 2.21877), (48.83961, 2.19702),
	(48.83612, 2.18306), (48.83166, 2.14879), (48.83243, 2.1392),
	(48.83964, 2.11881), (48.84222, 2.09905), (48.84196, 2.09547),
	(48.8128, 2.05831), (48.79852, 2.04373), (48.79692, 2.04008),
	(48.79496, 2.04106), (48.79544, 2.04537), (48.79446, 2.04934),
	(48.78763, 2.06262), (48.77713, 2.06017), (48.77094, 2.05524),
	(48.76883, 2.06241), (48.765, 2.06846)
]

def bench(title, method, arg):
	"""
	Run the comparison and print the results.
	"""
	is_tty = sys.stdout.isatty()
	if is_tty:
		print(f'\n\033[7;1m{title.center(75)}\033[27;0m\n')
	else:
		print(f'\n# {title}\n')

	fp_ms = 1_000 * timeit.timeit(lambda: getattr(fast_polyline, method)(arg), number=10_000)
	p_ms = 1_000 * timeit.timeit(lambda: getattr(polyline, method)(arg), number=10_000)

	if not is_tty:
		print('```')
	print(f'fast_polyline:\t{fp_ms:.5}ms')
	print(f'polyline:\t{p_ms:.5}ms')
	print(f'fast_polyline is {p_ms / fp_ms:.3} times faster.')
	if not is_tty:
		print('```')

bench('encode', 'encode', POINTS)
bench('decode', 'decode', POLYLINE)


src/fast_polyline_ext.c
#define PY_SSIZE_T_CLEAN
#include <Python.h>

/* # Resources
 *
 * [Python C-API reference](https://docs.python.org/3/c-api/index.html).
 */

/*
 * TODO: readme (.rst???)
 * TODO: test
 *   - check that error messages are set.
 * TODO: offer merging with the existing polyline project
 *   https://github.com/frederickjansen/polyline/issues/14
 * TODO: choose final naming (- or _ depending on how they do in python)
 */

// An encoded number can have at most _precision_ characters. However,
// it seems like we have a fuzzy behavior on low precisions. Hence a guard
// for those lower precisions.
#define MAX_ENCODED_CHUNKS(precision) (precision < 6 ? 6 : precision)

// Already smaller than a sand grain. https://xkcd.com/2170/
// In fact, at 15 precision, we can be sure there will be precision loss
// in the C/Ruby value conversion. And 14 precision may be an edge case
// since we work with signed values.
#define MAX_PRECISION 13

#ifdef DEBUG
#define dbg(...) printf(__VA_ARGS__)
#else
#define dbg(...)
#endif

typedef unsigned int uint;

// Speed up the code a bit by not calling pow(10.0, precision) since it is
// always 10^precision. Convert that into a lookup instead of doing the
// calculation.
// The precision parameter is already checked in +_get_precision()+ to ensure
// that it is within the range.
static inline double _fast_pow10(uint precision) {
	const double lookup[MAX_PRECISION + 1] = {
	    1.0,    10.0,   1.0e+2, 1.0e+3,  1.0e+4,  1.0e+5,  1.0e+6,
	    1.0e+7, 1.0e+8, 1.0e+9, 1.0e+10, 1.0e+11, 1.0e+12, 1.0e+13};
	return lookup[precision];
}

// Ensure that precision is between 0 and 13, set error string
// otherwise.
static inline int _check_precision(int precision) {
	if (precision > MAX_PRECISION) {
		PyErr_SetString(PyExc_ValueError,
		                "precision too high (https://xkcd.com/2170/)");
		return 0;
	}
	if (precision < 0) {
		PyErr_SetString(PyExc_ValueError,
		                "negative precision doesn't make sense");
		return 0;
	}
	return 1;
}

// Wrapper around PyArg_ParseTuple that sets an error message.
static inline int _check_args(PyObject *args, const char *fmt, ...) {
	// Passing through variadic arguments is not that easy in C.
	// Here's some information on the topic:
	// https://c-faq.com/varargs/handoff.html
	va_list argp;
	va_start(argp, fmt);
	int ret = PyArg_VaParse(args, fmt, argp);
	va_end(argp);
	if (!ret) PyErr_SetString(PyExc_TypeError, "invalid arguments");
	return ret;
}

// polyline.decode(polyline, precision)
static PyObject *polyline_decode(PyObject *self, PyObject *args) {
	const char *polyline;
	int precision;
	if (!_check_args(args, "si", &polyline, &precision)) return NULL;
	if (!_check_precision(precision)) return NULL;
	double precision_value = _fast_pow10(precision);
	PyObject *ary = PyList_New(0);
	// Helps keeping track of whether we are computing lat (0) or lng (1).
	uint8_t index = 0;
	size_t shift = 0;
	int64_t delta = 0;
	PyObject *pair[2];
	double latlng[2] = {0.0, 0.0};
	// Loops until end of string nul character is encountered.
	while (*polyline) {
		int64_t chunk = *polyline++;

		if (chunk < 63 || chunk > 126) {
			PyErr_Format(PyExc_ValueError, "invalid character '%c'",
			             (char)chunk);
			return NULL;  // TODO: ensure that the objects we created will be
			              // gc'd
		}

		chunk -= 63;
		delta = delta | ((chunk & ~0x20) << shift);
		shift += 5;

		if (!(chunk & 0x20)) {
			delta = (delta & 1) ? ~(delta >> 1) : (delta >> 1);
			latlng[index] += delta;
			pair[index] =
			    PyFloat_FromDouble((double)latlng[index] / precision_value);
			// When both coordinates are parsed, we can push those to the result
			// ary.
			if (index) PyList_Append(ary, PyTuple_Pack(2, pair[0], pair[1]));
			// Reinitilize since we are done for current coordinate.
			index = 1 - index;
			delta = 0;
			shift = 0;
		}
	}
	return ary;
}

static inline uint8_t _polyline_encode_number(char *chunks, int64_t number) {
	dbg("_polyline_encode_number(\"%s\", %lli)\n", chunks, number);
	number = number < 0 ? ~(number << 1) : (number << 1);
	uint8_t i = 0;
	while (number >= 0x20) {
		uint8_t chunk = number & 0x1f;
		chunks[i++] = (0x20 | chunk) + 63;
		number = number >> 5;
	}
	chunks[i++] = number + 63;
	dbg("%u encoded chunks\n", i);
	dbg("chunks: '%s'\n", chunks);
	dbg("/_polyline_encode_number\n");
	return i;
}

// polyline.encode(points, precision=5)
static PyObject *polyline_encode(PyObject *self, PyObject *args) {
	PyObject *ary;
	// Dynamically checked for [0; 13] inclusion.
	int precision;
	if (!_check_args(args, "Oi", &ary, &precision)) return NULL;
	if (!_check_precision(precision)) return NULL;
	double precision_value = _fast_pow10(precision);
	if (!PyList_Check(ary)) {
		PyErr_SetString(PyExc_TypeError, "points must be a list");
		return NULL;
	}

	size_t len = PyList_Size(ary);
	uint64_t i;
	PyObject *current_pair;
	int64_t prev_pair[2] = {0, 0};

	size_t alloc_size = MAX_ENCODED_CHUNKS(precision) * 2 * len;
	dbg("allocated size: %u * 2 * %lu = %lu\n", MAX_ENCODED_CHUNKS(precision),
	    len, alloc_size);
	char *chunks = malloc(alloc_size * sizeof(char));
	size_t chunks_index = 0;
	dbg("ary.len: %lu\n", len);
	for (i = 0; i < len; i++) {
		current_pair = PyList_GetItem(ary, i);
		uint8_t j;
		if (PyList_Check(current_pair))
			current_pair = PyList_AsTuple(current_pair);
		if (!PyTuple_Check(current_pair)) goto points_error;
		if (PyTuple_Size(current_pair) != 2) goto points_error;

		for (j = 0; j < 2; j++) {
			PyObject *current_coord = PyTuple_GetItem(current_pair, j);
			int is_float = PyFloat_Check(current_coord);
			int is_long = PyLong_Check(current_coord);
			if (!is_float && !is_long) goto points_error;

			double coord = is_float ? PyFloat_AsDouble(current_coord)
			                        : PyLong_AsDouble(current_coord);
			if (-180.0 > coord || coord > 180.0) goto points_error;

			int64_t rounded_value = round(coord * precision_value);
			int64_t delta = rounded_value - prev_pair[j];
			prev_pair[j] = rounded_value;
			// We pass a pointer to the current chunk that need to be filled.
			// Doing so avoid having to copy the string every single iteration.
			chunks_index += _polyline_encode_number(
			    chunks + chunks_index * sizeof(char), delta);
		}
	}

	assert(chunks_index < alloc_size);
	dbg("final chunks_index: %lu\n", chunks_index);
	PyObject *polyline = PyUnicode_FromStringAndSize(chunks, chunks_index);
	free(chunks);
	return polyline;

points_error:
	PyErr_SetString(PyExc_TypeError,
	                "points must be a list of (lat, lng) pairs");
	free(chunks);
	return NULL;
}

static PyMethodDef methods[] = {{"decode", polyline_decode, METH_VARARGS, ""},
                                {"encode", polyline_encode, METH_VARARGS, ""},
                                {NULL, NULL, 0, NULL}};

static struct PyModuleDef module = {PyModuleDef_HEAD_INIT, "fast_polyline_ext",
                                    "", -1, methods};
PyMODINIT_FUNC PyInit_fast_polyline_ext(void) {
	return PyModule_Create(&module);
}


src/fast_polyline.py
"""
Fast & Easy Google Polylines

Implementation of the `Google polyline algorithm
<https://code.google.com/apis/maps/documentation/utilities/polylinealgorithm.html>`_

.. code-block:: language-python
    fast_polyline.encode([[38.5, -120.2], [40.7, -120.95], [43.252, -126.453]])
    # => '_p~iF~ps|U_ulLnnqC_mqNvxq`@'

    fast_polyline.decode('_p~iF~ps|U_ulLnnqC_mqNvxq`@')
    # => [[38.5, -120.2], [40.7, -120.95], [43.252, -126.453]]

You can set an arbitrary precision for your coordinates to be
encoded/decoded. It may be from 1 to 13 decimal digits. However,
13 may be `too much
<https://www.explainxkcd.com/wiki/index.php/2170:_Coordinate_Precision>`_.

.. image:: https://imgs.xkcd.com/comics/coordinate_precision.png
        :alt: Coordinate Precision explained by XKC
"""
import fast_polyline_ext

def encode(coordinates, precision=5):
	"""
	Encode a list of (lat, lng) coordinates
	into a Polyline
	"""
	return fast_polyline_ext.encode(coordinates, precision)

def decode(polyline, precision=5):
	"""
	Decode a Polyline into a list of (lat, lng)
	coordinate tuples
	"""
	return fast_polyline_ext.decode(polyline, precision)

__all__ = ['encode', 'decode']
